ffi/src/lib.rs

```diff
@@ -178,3 +555,131 @@ pub unsafe extern "C" fn store_sync(store: *mut Store, user_uuid: *const c_char,
+    let store = &mut*store;
+    let kw = kw_from_string(c_char_to_string(attribute));
+    let value = kw_from_string(c_char_to_string(value));
+    let is_valid = store.conn().current_schema().get_entid(&value);
+    if is_valid.is_none() {
+        return Box::into_raw(Box::new(ExternResult { ok: std::ptr::null_mut(), err: string_to_c_char(format!("Unknown attribute {:?}", value)) }));
+    }
+    let kw_entid = is_valid.unwrap();
+    let res = store.add_value_for_attribute(KnownEntid(entid), kw, TypedValue::Ref(kw_entid.into()));
+    Box::into_raw(Box::new(res.into()))
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn store_set_boolean_for_attribute_on_entid(store: *mut Store, entid: Entid, attribute: *const c_char, value: bool) -> *mut ExternResult {
+    let store = &mut*store;
+    add_value_for_attribute(store, KnownEntid(entid), c_char_to_string(attribute), value)
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn store_set_double_for_attribute_on_entid(store: *mut Store, entid: Entid, attribute: *const c_char, value: f64) -> *mut ExternResult {
+    let store = &mut*store;
+    add_value_for_attribute(store, KnownEntid(entid), c_char_to_string(attribute), value)
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn store_set_timestamp_for_attribute_on_entid(store: *mut Store, entid: Entid, attribute: *const c_char, value: time_t) -> *mut ExternResult {
+    let store = &mut*store;
+    let kw = kw_from_string(c_char_to_string(attribute));
+    let res = store.add_value_for_attribute(KnownEntid(entid), kw, TypedValue::instant(value as i64));
+    Box::into_raw(Box::new(res.into()))
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn store_set_string_for_attribute_on_entid(store: *mut Store, entid: Entid, attribute: *const c_char, value: *const c_char) -> *mut ExternResult {
+    let store = &mut*store;
+    add_value_for_attribute(store, KnownEntid(entid), c_char_to_string(attribute), c_char_to_string(value))
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn store_set_uuid_for_attribute_on_entid(store: *mut Store, entid: Entid, attribute: *const c_char, value: *const c_char) -> *mut ExternResult {
+    let store = &mut*store;
+    let uuid = Uuid::parse_str(&c_char_to_string(value)).expect("valid uuid");
+    add_value_for_attribute(store, KnownEntid(entid), c_char_to_string(attribute), uuid)
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn destroy(obj: *mut c_void) {
+    if !obj.is_null() {
+        let obj_to_release = Box::from_raw(obj);
+        println!("object to release {:?}", obj_to_release);
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn query_builder_destroy(obj: *mut QueryBuilder) {
+    if !obj.is_null() {
+        let _ = Box::from_raw(obj);
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn store_destroy(obj: *mut Store) {
+    if !obj.is_null() {
+        let _ = Box::from_raw(obj);
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn typed_value_destroy(obj: *mut TypedValue) {
+    if !obj.is_null() {
```

Comment:
```
macro_rules! define_destructor {
    ( $name:ident, $t:ty ) => {{
        #[no_mangle]
        pub unsafe extern "C" fn $name(obj: *mut $t) {
            if !obj.is_null() { let _ = Box::from_raw(obj); }
        }
    }}
}
```

?

Reply:
ooh, I didn't know you could do that.

Reply:
Oh, weird. Adding that function means that _no_ FFI functions can be found from Swift. I guess another lesson - no macros in FFI!